cupy/_sorting/search.py

```diff
@@ -212,6 +212,32 @@ def where(condition, x=None, y=None):
     return _where_ufunc(condition.astype('?'), x, y)
 
 
+def argwhere(a):
+    """Return the indices of the elements that are non-zero.
+
+    Returns a (N, ndim) dimantional array containing the
+    indices of the non-zero elements. Where `N` is number of
+    non-zero elements and `ndim` is dimention of the given array.
+
+    Args:
+        a (cupy.ndarray): array
+
+    Returns:
+        cupy.array: Indices of elements that are non-zero.
+
+    .. seealso:: :func:`numpy.argwhere`
+
+    """
+    util.check_array(a, arg_name='a')
+    # promote 0d to 1d because argwhere doesn't work well with 0d
+    if a.ndim == 0:
+        a = cupy.shape_base.atleast_1d(a)
+        # then remove the added dimension
+        return argwhere(a)[:, :0]
+    a = cupy.asarray(cupy.nonzero(a), dtype=int)
```

Comment:
Actually I'd be tempted to do the opposite. Namely pull the `nonzero` code over here and then rewrite `nonzero` in terms of `argwhere`. Here's [how we did this in Dask for example]( https://github.com/dask/dask/blob/dfcc09c6eb48d59e391e593408c12181f6550f31/dask/array/routines.py#L1244-L1249 ).

Reply:
Current implementation causes redundant memory copy. How about using `_nonzero_kernel` in `_routines_indexing.pyx` directly?

Reply:
Do you mean I should implement `argwhere` in `_routines_indexing.pyx` as `nonzero` is implemented there?

Reply:
Do you mean implementing `argwhere` in  `_routines_indexing.pyx` like `nonzero` is implemented there?

Reply:
And I am sorry for the late response @jakirkham

Reply:
No worries. Not late at all 😄

Yeah that's my thinking 🙂

Would be tempted to just rename [`_ndarray_nonzero`]( https://github.com/cupy/cupy/blob/7cf5fef1f1f568b04180b20b789d2a08fdd2b9f0/cupy/core/_routines_indexing.pyx#L51 ) to `_ndarray_argwhere` and tweak [the return to be `dst`]( https://github.com/cupy/cupy/blob/7cf5fef1f1f568b04180b20b789d2a08fdd2b9f0/cupy/core/_routines_indexing.pyx#L76 ). Could still keep `_ndarray_nonzero` around as a wrapper that just calls `_ndarray_argwhere`. Should be straightforward to use `_ndarray_argwhere` here. If that makes sense 😉

Akifumi may have more thoughts 🙂